blocks/graph.py

```diff
@@ -204,10 +204,15 @@ def replace(self, replacements):
         # Sort `replacements` in topological order
         for node in apply_nodes_sorted:
             for input_ in node.inputs:
-                if input_ not in replacements:
-                    continue
-                replacement_keys_cur.append(input_)
-                replacement_vals_cur.append(replacements[input_])
+                if input_ in replacements:
```

Comment:
OH! I get it, I think. So you're saying that if a node to be replaced is an input of multiple nodes in the graph, its replacement will happen more than once which will affect the correctness of the computation.

Here's a minimal example that reproduces the issue:
```python
from theano import tensor
from theano.printing import debugprint
from blocks.graph import ComputationGraph

x = tensor.scalar('x')
y = 2 * x
z = x + 1

cg = ComputationGraph([y, z])
debugprint(cg.outputs[0])
cg = cg.replace({x: 0.5 * x})
print('----------')
debugprint(cg.outputs[0])
print('----------')
print cg.outputs[0].eval({x: 1.0}), cg.outputs[1].eval({x: 1.0})
```

The expected output for `y` and `z` is 1 and 1.5, respectively, but the output produced by the code snippet above is 0.5 and 1.25, which corresponds to the replacement being applied twice on `x`, as shown in the `debugprint` output.

@rizar I'll make a pull request for that right away.